src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs

```diff
@@ -126,23 +125,27 @@ protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(string code,
 
         protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri)
         {
-            var oauthstate = Options.StateDataFormat.Protect(properties);
-            if (Options.AuthorizationEndpoint != WeixinAuthenticationDefaults.AuthorizationEndpoint)
+            var state = Options.StateDataFormat.Protect(properties);
+            if (!IsWeixinAuthorizationEndpointInUse())
             {
                 //Store state in redirectUri when authorizing Wechat Web pages to prevent "too long state parameters" error
-                redirectUri = redirectUri.Contains("?") ? $"{redirectUri}&{nameof(oauthstate)}={oauthstate}" : $"{redirectUri}?{nameof(oauthstate)}={oauthstate}";
-                oauthstate = "#wechat_redirect";//The Parameters Necessary for Web Authorization of Wechat
+                redirectUri = QueryHelpers.AddQueryString(redirectUri, OauthState, state);
+                state = "#wechat_redirect";//The Parameters Necessary for Web Authorization of Wechat
```

Comment:
So just to be clear, Weixin expects the `state` parameter value to always end with `#wechat_redirect`, or is what you're meaning to do is put a hash at the end of the whole URL?

I suspect it's the latter, otherwise you'd have to URI encode the `#`.

For that, you'd just do this:

```csharp
return QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, new Dictionary<string, string>
            {
                ["appid"] = Options.ClientId,
                ["scope"] = FormatScope(),
                ["response_type"] = "code",
                ["redirect_uri"] = redirectUri,
                ["state"] = oauthstate
                ["state"] = state
            }) + "#wechat_redirect";
```

Which could then probably be refactored to:

```csharp
var state = Options.StateDataFormat.Protect(properties);
var addRedirectHash = false;

if (!IsWeixinAuthorizationEndpointInUse())
{
    redirectUri = QueryHelpers.AddQueryString(redirectUri, OauthState, state);
    addRedirectHash = true;
}

redirectUri = QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, new Dictionary<string, string>
            {
                ["appid"] = Options.ClientId,
                ["scope"] = FormatScope(),
                ["response_type"] = "code",
                ["redirect_uri"] = redirectUri,
                ["state"] = oauthstate
                ["state"] = state
            });

if (addRedirectHash)
{
    // The parameters necessary for Web Authorization of Wechat
    redirectUri += "#wechat_redirect";
}

return redirectUri;
```